text/0000-libsyntax2.0.md

```diff
@@ -0,0 +1,435 @@
+data structure which used in [IntelliJ] based IDEs and in the [Kotlin]
+compiler.
+
+
+[PSI]: http://www.jetbrains.org/intellij/sdk/docs/reference_guide/custom_language_support/implementing_parser_and_psi.html
+[IntelliJ]: https://github.com/JetBrains/intellij-community/
+[Kotlin]: https://kotlinlang.org/
+
+
+## Untyped Tree
+
+The main idea is to store the minimal amount of information in the
+tree itself, and instead lean heavily on the source code for the
+actual data about identifier names, constant values etc.
+
+All nodes in the tree are of the same type and store a constant for
+the syntactic category of the element and a range in the source code.
+
+Here is a minimal implementation of this data structure with some Rust
+syntactic categories
+
+
+```rust
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct NodeKind(u16);
+
+pub struct File {
+	text: String,
+	nodes: Vec<NodeData>,
+}
+
+struct NodeData {
+	kind: NodeKind,
+	range: (u32, u32),
+	parent: Option<u32>,
+	first_child: Option<u32>,
+	next_sibling: Option<u32>,
+}
+
+#[derive(Clone, Copy)]
+pub struct Node<'f> {
+	file: &'f File,
+	idx: u32,
+}
+
+pub struct Children<'f> {
+	next: Option<Node<'f>>,
+}
+
+impl File {
+	pub fn root<'f>(&'f self) -> Node<'f> {
+		assert!(!self.nodes.is_empty());
+		Node { file: self, idx: 0 }
+	}
+}
+
+impl<'f> Node<'f> {
+	pub fn kind(&self) -> NodeKind {
+		self.data().kind
+	}
+
+	pub fn text(&self) -> &'f str {
+		let (start, end) = self.data().range;
+		&self.file.text[start as usize..end as usize]
+	}
+
+	pub fn parent(&self) -> Option<Node<'f>> {
+		self.as_node(self.data().parent)
+	}
+
+	pub fn children(&self) -> Children<'f> {
+		Children { next: self.as_node(self.data().first_child) }
+	}
+
+	fn data(&self) -> &'f NodeData {
+		&self.file.nodes[self.idx as usize]
+	}
+
+	fn as_node(&self, idx: Option<u32>) -> Option<Node<'f>> {
+		idx.map(|idx| Node { file: self.file, idx })
+	}
+}
+
+impl<'f> Iterator for Children<'f> {
+	type Item = Node<'f>;
+
+	fn next(&mut self) -> Option<Node<'f>> {
+		let next = self.next;
+		self.next = next.and_then(|node| node.as_node(node.data().next_sibling));
+		next
+	}
+}
+
+pub const ERROR: NodeKind = NodeKind(0);
+pub const WHITESPACE: NodeKind = NodeKind(1);
+pub const STRUCT_KW: NodeKind = NodeKind(2);
+pub const IDENT: NodeKind = NodeKind(3);
+pub const L_CURLY: NodeKind = NodeKind(4);
+pub const R_CURLY: NodeKind = NodeKind(5);
+pub const COLON: NodeKind = NodeKind(6);
+pub const COMMA: NodeKind = NodeKind(7);
+pub const AMP: NodeKind = NodeKind(8);
+pub const LINE_COMMENT: NodeKind = NodeKind(9);
+pub const FILE: NodeKind = NodeKind(10);
+pub const STRUCT_DEF: NodeKind = NodeKind(11);
+pub const FIELD_DEF: NodeKind = NodeKind(12);
+pub const TYPE_REF: NodeKind = NodeKind(13);
+```
+
+Here is a rust snippet and the corresponding parse tree:
+
+```rust
+struct Foo {
+	field1: u32,
+	&
+	// non-doc comment
+	field2:
+}
+```
+
+
+```
+FILE
```

Comment:
I'm wondering: How do we figure out the best way to interpret a partial tree? Should we rely on indentation as a hint? Are there well-known solutions here?

Reply:
Here's the array of tricks I am aware of:

1) Code is typically typed left-to-right, so it's possible to have "commit points" in grammar/parser. Here's an example from fall:

```
pub rule fn_def {
  'const'? 'unsafe'? linkage?
  'fn' <commit> ident
  type_parameters?
  value_parameters
  ret_type?
  where_clause?
  {block_expr | ';'}
}
```

This `<commit>` means that as soon parser sees `fn` keyword, it produces a function node, even if there are parsing errors after this point. Commit effectively makes the trailing part optional. In other words, one way to deal with partial parses is to treat certain prefixes of parses as parses. These commit points nicely mesh up with `ll`-ness of the parser: you commit just after the necessary lookahead. 

2) Code usually has block structure, so it makes sense, when parsing a block expression, first parse it approximately as a token tree, and then try to parse the internals of the block. That way, parse errors stay isolated to blocks.  Of course, you can do this with all kinds of lists and what not, as long as you can invent a robust covering grammar. This trick also makes incremental parsers more efficient (as changes are isolated to one block unless it's a change in the block structure itself) and allows one to lazily parse stuff. There are certain variations of this trick: for example, you can *lex* string literals as just *any* stuff between `"`, and then lex the contents of the literal with a second lexer, which properly deals with escape sequences. 

3) When parsing repeated constructs, like `item*`, a useful strategy of error recovery is, after each (partially) parsed item skip tokens until a token from `FIRST(item)` appears. ([example from fall](https://github.com/matklad/fall/blob/c49f376fc38c8ed2ed42d4eb27f057cda7139483/lang/rust/src/rust.fall#L109-L127))